### DISCUSSION TITLE
Unicode and transliteration

### DISCUSSION THREAD

**Discussion by Torinde** [dosbox-staging/dosbox-staging#2592] [Ideas]
Following #2548 by @FeralChild64 and looking forward to #1864 

A few characters can get better analogues `Host Unicode -> DOS`
- when DOS codepage is 3021, 872, 855:
CYRILLIC CAPITAL LETTER I WITH GRAVE **Ѝ -> Й** CYRILLIC CAPITAL LETTER SHORT I (currently it gives "?")
CYRILLIC SMALL LETTER I WITH GRAVE **ѝ -> й** CYRILLIC SMALL LETTER SHORT I (currently it gives "?")
CYRILLIC CAPITAL LETTER BYELORUSSIAN-UKRAINIAN I (Windows keyboard layout output [for Roman number 1](https://learn.microsoft.com/en-us/globalization/keyboards/kbdbu) `Shift -` = `І`) -> LATIN CAPITAL LETTER I (instead of the current DOSbox output of `И` CYRILLIC CAPITAL LETTER I)
- when DOS codepage is 437 (or similarly restricted)
CYRILLIC CAPITAL LETTER I WITH GRAVE **Ѝ -> ì** LATIN SMALL LETTER I WITH GRAVE
CYRILLIC SMALL LETTER I WITH GRAVE **ѝ -> ì** LATIN SMALL LETTER I WITH GRAVE
- when DOS codepage has 238 (EEhex) as Greek "[Epsilon](https://en.wikipedia.org/wiki/Epsilon) Small" [U+03B5, ε] or [element-of](https://en.wikipedia.org/wiki/Element_(mathematics)) sign [U+2208, ∈]:
EURO **€ -> ε/∈**  per [Wikipedia](https://en.wikipedia.org/wiki/Code_page_437) "It was later unofficially used as the [euro sign](https://en.wikipedia.org/wiki/Euro_sign) [U+20AC, €]"
- when DOS codepage has **section sign §** - don't replace it with `#` (except when using the most limited set of "Printable ASCII characters"). Currently with CP437 it gives `#`?
- **Ruble sign ₽** currently gives "?" - I suggest using Latin "P" instead (or Cyrillic "Р")

Transliteration - as explained at [Informal romanizations of Cyrillic](https://en.wikipedia.org/wiki/Informal_romanizations_of_Cyrillic#Translit_in_Bulgaria), transliteration depends not only on the source codepage/symbols, but also on the language - current one (e.g. Cyrillic letters representation via CP437) fits some languages, but is unusable for others. Would be nice to have check the source "locale"/region/language and adjust the mapping accordingly.

**Comment by FeralChild64** [comment]
1. Regarding BYELORUSSIAN-UKRAINIAN I - OK, I can change it
2. Regarding EURO SIGN - OK, I can add aliases
3. Regarding SECTION SIGN - we cannot map it to ASCII code 0x15, as this is also a control code (NAK - network acknowledgement), having this in a string either won't work or would be risky
4. Regarding RUBLE SIGN and CYRILLIC LETTER I WITH GRAVE - do you have any use-case for them (like, are you preparing some translation which needs these signs)? Currently no code page supported by _DOSBox Staging_ contains these characters - if there is a real need I can add them, however they will always be mapped to some fallback character 
5. Regarding language-dependent transliteration - it can be added, but:
- somebody has to provide me with rules for the given language
- after 0.81.0 release; currently the Unicode engine is not aware of the country/language being set (it only knows the display code page), but it will have to change one day nevertheless for proper sorting support and to export collation table to DOS software

**Reply by Torinde** [reply]
4. One use-case is for pasting text from Host into DOSbox as mentioned in #1864. Another use-case is for writing into the `autoexec` section of .conf. Another is for host-created filenames that have that such character. And yes, the fallback character should be й as mentioned... Actually CP437 ù (0x97) looks similar, but I haven't seen it in any of the Cyrillic DOS codepages. All characters that I mentioned are present in Windows keyboard layouts, e.g. user can type those directly, without Character map or numpad codes or anything like that.
5. OK, so I'll prepare a romanization table after 0.81.0. When you say "aware of the country/language being set" - do you mean "country in DOSbox" or "country on the Host"? I'm thinking here as well of "country on the Host", e.g. when writing in .conf or even better - copy-pasting from Host into DOSbox - if DOSbox codepage is 437 and I'm pasting Unicode Cyrillic characters - those to be transliterated to a readable/useful words, depending on the language of the Host (Source side).

**Comment by FeralChild64** [comment]
4. There are tens, or even hundreds of such Latin, Greek, Cyrillic, or Georgian characters. One day I intend to fill-in all the holes in Unicode blocks which are of interest to us, but not now. BTW, clipboard sharing is not yet implemented (I intend to implement it in the `EDIT.COM` replacement I am currently slowly working on), and accented or non-Latin characters in file names are not implemented yet either (believe me, it requires a lot of work to do this properly).

5. I am talking about `language =` and `country =` settings from the DOSBox Staging configuration file. Country/language on host might be used to set these values automatically, but this is platform-dependent code. Just keep in mind, that any transliteration has to transform a single glyph into a single glyph, transliterating `Æ` into `AE` is not allowed.

-----

BTW, you can experiment with Unicode by yourself, without even recompiling the emulator - go to `contrib/resources/mapping` (probably just `resources/mapping` after installing on Windows) and edit the files:
- `ASCII.TXT` contains the last-resort fallback, to 7-bit ASCII characters (0x20-0x7F), has to be filled-in for every single code point we support
- `CAPITAL_SMALL.TXT` contains uppercase/lowercase character mapping - this is currently not yet utilized anywhere, but will be needed for future functionalities
- `DECOMPOSITION.TXT` tells how to decompose characters to normalization form KD - has to be filled in only for glyphs which can be decomposed into two or more code points
- `MAIN.TXT` is the main code page definition file; also contains aliases, like _you can replace CENT SIGN with LATIN SMALL LETTER C WITH STROKE_

**Reply by Torinde** [reply]
4. Sure, the one letter I mentioned is from the keyboard layouts I checked. Of course there will be more... are there already a correspondence tables somewhere? Or an automatic way to do that? If the process is manual I'm willing to help you... although I'm not sure how useful help is from a person not speaking the language...

Thanks for TXT explanation - I'll try those...

**Reply by Torinde** [reply]
@FeralChild64, about 5. Country/language on host might be used to set these values automatically, but this is platform-dependent code.

Staging already detects the Host regional format (or the non-English preferred language) on Windows - see portable mode with empty dosbox-staging.conf file:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/c5cc8460-5eb1-4a3e-83f4-54809b7ae7ae)
![image](https://github.com/dosbox-staging/dosbox-staging/assets/97228894/42684d36-16ec-425e-ab5f-f4a61c8cbc7f)

Is that working only on Windows and not macOS and Linux?

If it works cross-platform, then can it be used to trigger [the romanization](https://github.com/dosbox-staging/dosbox-staging/blob/91e65ff30a109de47ebd743ff2a6d0a466b6214e/contrib/resources/mapping/ASCII.TXT#L822)?

A user setting to choose the romanization may also be nice (e.g. to switch to the other list without changing Host settings).

**Reply by rderooy** [reply]
On Linux this uses the "LANG" environment variable, which encodes both the language and country. e.g. "de_CH.UTF-8" means German language for Switzerland, where instead "fr_FR.UTF-8" means French language for France.

Unfortunately this does not tell you the host keyboard layout, which is guessed based on the language and country, and can therefore be wrong.

**Reply by FeralChild64** [reply]
> Is that working only on Windows and not macOS and Linux?

I haven’t investigated it in detail, but POSIX implementation is definitely lacking.

> If it works cross-platform, then can it be used to trigger [the romanization](https://github.com/dosbox-staging/dosbox-staging/blob/91e65ff30a109de47ebd743ff2a6d0a466b6214e/contrib/resources/mapping/ASCII.TXT#L822)?

> A user setting to choose the romanization may also be nice (e.g. to switch to the other list without changing Host settings).

Right now there is no support for country-specific romanization - and currently there is no use case for this. There is no clipboard sharing implemented yet, no Bulgarian translation available, etc. Zero gain for the end user.

**Comment by FeralChild64** [comment]
There is no automatic way to do this. Files contains comments and explanations. For the list of Unicode blocks and characters, I like the page https://www.compart.com/en/unicode/block - just note, that we don't want to support everything, DOS cannot handle hieroglyphs :) In `ASCII.TXT` you will find the block names we are currently using. Do not add combining marks.

**Comment by FeralChild64** [comment]
Example: CYRILLIC CAPITAL LETTER I WITH GRAVE - https://www.compart.com/en/unicode/U+040D

1. It is U+040D, the last resort fallback would be letter `I`, so in ASCII.TXT we need a line (just after line with `0x040c`):
```0x040d I   #CYRILLIC CAPITAL LETTER I WITH GRAVE```
2. The web page tells us the character can be decomposed into [И (U+0418)](https://www.compart.com/en/unicode/U+0418) - [◌̀ (U+0300)](https://www.compart.com/en/unicode/U+0300), so we need to update DECOMPOSITION.TXT and add a line:
```0x040d  0x0418 0x0300         #CYRILLIC CAPITAL LETTER I WITH GRAVE```
3. We need to check if 0x0418 is already supported - it is, we can find this character in ASCII.TXT (if it wouldn't, we would have to do all these steps for U+0418 as well). We do not add 0x0300, as this is a combining mark, not a letter.
4. We want an alias, so that CYRILLIC LETTER I (U+0418) can be used instead of CYRILLIC LETTER I WITH GRAVE (U+040D) if the code page contains the first character, but not the letter - this line has to be added to MAIN.TXT:
```ALIAS 0x040d 0x0418 #CYRILLIC CAPITAL LETTER I WITH GRAVE```
You can add more aliases, first applicable one will be used.
5. We need to specify, that CYRILLIC CAPITAL LETTER I WITH GRAVE (U+040D) is an uppercase version of CYRILLIC SMALL LETTER I WITH GRAVE (U+045D) - we need to add a line to CAPITAL_SMALL.TXT:
```0x040d 0x045d #CYRILLIC LETTER I WITH GRAVE```
6. And since CYRILLIC SMALL LETTER I WITH GRAVE is not supported yet, we need to do all the steps for this letter too.

**Reply by Torinde** [reply]
Thanks for the detailed breakdown - I'll try it with the TXT files.

Just one remark - the fallback letter is CYRILLIC CAPITAL LETTER SHORT I. If not present in the target codepage, then CYRILLIC LETTER I. If that is not present - then LATIN SMALL LETTER I WITH GRAVE, and finally if even that is not present - LATIN CAPITAL LETTER I

ѝ -> й -> и -> ì -> i
Ѝ -> Й -> И -> ì -> I

This letter is very seldom used as capital (e.g. only in a title with all capitals; a sentence can't begin with that letter). At least for codepages 3021, 872, 855. So, if there is some odd codepage that has ѝ or й, but neither Ѝ nor Й, then the small ѝ or й or be used in the place of Ѝ... 

If we can't have so many fallbacks, then the most important one is ѝ -> й

Anyway, I'll try editing the TXT files and will share here what I find as good addition to them.

**Comment by FeralChild64** [comment]
In MAIN.TXT you can declare as many fallbacks (aliases) as needed.
Fallback to LATIN CAPITAL LETTER I is the last-resort generic fallback, it should be specified in ASCII.TXT, but not in MAIN.TXT.

**Comment by Torinde** [comment]
OK, should I make a PR with the following?

DECOMPOSITION.TXT
```
0x040d  0x0418 0x0300         #CYRILLIC CAPITAL LETTER I WITH GRAVE
0x045d  0x0438 0x0300         #CYRILLIC SMALL LETTER I WITH GRAVE
0x0419  0x0418 0x0306         #CYRILLIC CAPITAL LETTER SHORT I
0x0439  0x0438 0x0306         #CYRILLIC SMALL LETTER SHORT I
```

CAPITAL_SMALL.TXT
`0x040d 0x045d #CYRILLIC LETTER I WITH GRAVE`

ASCII.TXT
```
0x040d I   #CYRILLIC CAPITAL LETTER I WITH GRAVE
0x045d i   #CYRILLIC SMALL LETTER I WITH GRAVE
```

MAIN.TXT
```
ALIAS 0x040d 0x0419 #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x045d 0x0439 #CYRILLIC SMALL LETTER I WITH GRAVE
ALIAS 0x040d 0x0418 #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x045d 0x0438 #CYRILLIC SMALL LETTER I WITH GRAVE

ALIAS 0x040d 0x045d #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x040d 0x0439 #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x040d 0x0438 #CYRILLIC CAPITAL LETTER I WITH GRAVE

ALIAS 0x040d 0x00cc #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x040d 0x00ec #CYRILLIC CAPITAL LETTER I WITH GRAVE
ALIAS 0x045d 0x00ec #CYRILLIC SMALL LETTER I WITH GRAVE
```

In addition to previous discussion - CYRILLIC LETTER SHORT I is also decomposable and I added that here.

Below is the romanization for source codepage 3021 - differences from default (Russian?) ASCII.TXT - to be added after 0.81/when language/country can be detected? (currently I tested it by changing the TXT files)
```
0x0416 J   #CYRILLIC CAPITAL LETTER ZHE
0x0419 Y   #CYRILLIC CAPITAL LETTER SHORT I
0x0426 C   #CYRILLIC CAPITAL LETTER TSE
0x0427 C   #CYRILLIC CAPITAL LETTER CHE
0x0428 W   #CYRILLIC CAPITAL LETTER SHA
0x0429 W   #CYRILLIC CAPITAL LETTER SHCHA
0x042a X   #CYRILLIC CAPITAL LETTER HARD SIGN
0x042c Y   #CYRILLIC CAPITAL LETTER SOFT SIGN
0x042e U   #CYRILLIC CAPITAL LETTER YU
0x042f Q   #CYRILLIC CAPITAL LETTER YA
0x0436 j   #CYRILLIC SMALL LETTER ZHE
0x0439 y   #CYRILLIC SMALL LETTER SHORT I
0x0446 c   #CYRILLIC SMALL LETTER TSE
0x0447 c   #CYRILLIC SMALL LETTER CHE
0x0448 w   #CYRILLIC SMALL LETTER SHA
0x0449 w   #CYRILLIC SMALL LETTER SHCHA
0x044a x   #CYRILLIC SMALL LETTER HARD SIGN
0x044c y   #CYRILLIC SMALL LETTER SOFT SIGN
0x044e u   #CYRILLIC SMALL LETTER YU
0x044f q   #CYRILLIC SMALL LETTER YA
```

**Reply by FeralChild64** [reply]
OK, you can create a pull request. Regarding this romanization - it would be best to put a comment in ASCII.TXT or MAIN.TXT ('TODO'), that for Bulgarian (am I right?) language this would be a better transliteration.
BTW. There is no 'source code page' - the input Unicode string does not contain such information, we only know the code page set inside the DOSBox.

**Comment by Torinde** [comment]
While testing the above I found out it seems DOSbox ECHO doesn't support caret sign to print special characters like pipe and others:
`ECHO ^|` to result in `|`, `ECHO ^>`, etc.
FreeDOS, CMD.EXE and others support that (MS-DOS doesn't), so I find it useful if DOSbox also does that.

**Comment by Torinde** [comment]
Done in #2621
